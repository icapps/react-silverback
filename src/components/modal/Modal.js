import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '../index';
import './modal.scss';

class Modal extends React.Component {
  constructor() {
    super();
    this.state = {
      showModal: false,
    };
  }

  componentDidMount() {
    this._ismounted = true;
  }

  componentWillUnmount() {
    this._ismounted = false;
  }

  toggleModal = () => {
    this.setState({ showModal: !this.state.showModal });
  };

  handleModalButton = () => {
    this.props.handleModalButton();
    this.toggleModal();
  };

  handleClosingButton = () => {
    this.toggleModal();
  };

  handlePrimaryButton = async () => {
    const result = await this.props.handlePrimaryButton();
    if (result && this._ismounted) {
      this.toggleModal();
    }
  };

  handleSecondaryButton = () => {
    this.props.handleSecondaryButton();
    this.toggleModal();
  };

  render() {
    const { props } = this;
    return (
      <React.Fragment>
        <button
          type="button"
          className={`btn ${props.modalButtonClassName}`}
          data-toggle="modal"
          data-target={`#${props.id}`}
          onClick={this.handleModalButton}
        >
          {props.icon && <img src={props.icon} alt="" />}
          {props.modalButtonText}
        </button>
        <div
          className={`modal fade ${this.state.showModal ? 'show' : ''}`}
          id={props.id}
          tabIndex="-1"
          role="dialog"
          aria-labelledby={props.id}
          aria-hidden="true"
        >
          <div className="modal-dialog modal-dialog-centered" role="document">
            <div className="modal-content">
              {props.hasHeader && (
                <div className="modal-header">
                  <h5 className="modal-title">{props.title}</h5>
                  <button
                    type="button"
                    className="close"
                    data-dismiss="modal"
                    aria-label="Close"
                    onClick={this.handleClosingButton}
                  >
                    <span aria-hidden="true">&times;</span>
                  </button>
                </div>
              )}
              <div className="modal-body">{props.children}</div>
              <div className="modal-footer">
                <button
                  type="button"
                  className={`btn ${props.secondaryButtonClassName}`}
                  data-dismiss="modal"
                  onClick={this.handleSecondaryButton}
                >
                  {props.secondaryButtonText}
                </button>
                <button
                  type="button"
                  className={`btn ${props.primaryButtonClassName}`}
                  data-dismiss="modal"
                  onClick={this.handlePrimaryButton}
                  disabled={props.isPending}
                >
                  {props.isPending ? (
                    <Spinner hasContainer={false} spinnerClassName={'button-spinner'} />
                  ) : (
                    props.primaryButtonText
                  )}
                </button>
              </div>
            </div>
          </div>
        </div>
        {this.state.showModal && <div className="modal-backdrop fade show"></div>}
      </React.Fragment>
    );
  }
}

Modal.propTypes = {
  id: PropTypes.string.isRequired,
  hasHeader: PropTypes.bool,
  title: PropTypes.string,
  modalButtonClassName: PropTypes.string,
  modalButtonText: PropTypes.string.isRequired,
  handleModalButton: PropTypes.func,
  secondaryButtonClassName: PropTypes.string,
  secondaryButtonText: PropTypes.string.isRequired,
  handleSecondaryButton: PropTypes.func,
  primaryButtonClassName: PropTypes.string,
  primaryButtonText: PropTypes.string.isRequired,
  handlePrimaryButton: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  icon: PropTypes.string,
  isPending: PropTypes.bool,
};

Modal.defaultProps = {
  modalButtonClassName: 'btn-light',
  hasHeader: false,
  title: '',
  secondaryButtonClassName: 'btn-light',
  primaryButtonClassName: 'btn-light',
  icon: null,
  isPending: false,
  handleModalButton: () => {},
  handleSecondaryButton: () => {},
};

export default Modal;
