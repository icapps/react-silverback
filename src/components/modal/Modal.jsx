import React from 'react';
import PropTypes from 'prop-types';
import './modal.css';

class Modal extends React.Component {
  constructor() {
    super();
    this.state = {
      showModal: false,
    };
  }

  toggleModal = () => {
    this.setState({ showModal: !this.state.showModal });
  }

  handleClosingButton = () => {
    this.toggleModal();
  }

  handlePrimaryButton = () => {
    this.props.handlePrimaryButton();
    this.toggleModal();
  }

  handleSecondaryButton = () => {
    this.props.handleSecondaryButton();
    this.toggleModal();
  }

  render() {
    const { props } = this;

    return (
      <React.Fragment>
        <button type="button" className={`btn ${props.modalButtonClassName}`} data-toggle="modal" data-target={`#${props.id}`} onClick={this.toggleModal}>
          {props.modalButtonText}
        </button>

        <div className={`modal fade ${this.state.showModal ? 'show' : ''}`} id={props.id} tabIndex="-1" role="dialog" aria-labelledby={props.id} aria-hidden="true">
          <div className="modal-dialog modal-dialog-centered" role="document">
            <div className="modal-content">
              {props.hasHeader && (
                <div className="modal-header">
                  <h5 className="modal-title">{props.title}</h5>
                  <button type="button" className="close" data-dismiss="modal" aria-label="Close" onClick={this.handleClosingButton}>
                    <span aria-hidden="true">&times;</span>
                  </button>
                </div>
              )}
              <div className="modal-body">
                {props.children}
              </div>
              <div className="modal-footer">
                <button type="button" className={`btn ${props.secondaryButtonClassName}`} data-dismiss="modal" onClick={this.handleSecondaryButton}>{props.secondaryButtonText}</button>
                <button type="button" className={`btn ${props.primaryButtonClassName}`} data-dismiss="modal" onClick={this.handlePrimaryButton}>{props.primaryButtonText}</button>
              </div>
            </div>
          </div>
        </div>
        {this.state.showModal && <div className="modal-backdrop fade show"></div>}
      </React.Fragment>
    );
  }
};

Modal.propTypes = {
  id: PropTypes.string.isRequired,
  modalButtonClassName: PropTypes.string,
  modalButtonText: PropTypes.string.isRequired,
  hasHeader: PropTypes.bool,
  title: PropTypes.string,
  secondaryButtonClassName: PropTypes.string,
  handleSecondaryButton: PropTypes.func.isRequired,
  secondaryButtonText: PropTypes.string.isRequired,
  primaryButtonClassName: PropTypes.string,
  handlePrimaryButton: PropTypes.func.isRequired,
  primaryButtonText: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

Modal.defaultProps = {
  modalButtonClassName: 'btn-light',
  hasHeader: false,
  title: '',
  secondaryButtonClassName: 'btn-light',
  primaryButtonClassName: 'btn-light',
};

export default Modal;
