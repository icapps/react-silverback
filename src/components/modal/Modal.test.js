import React from 'react';
import { shallow, configure } from 'enzyme';
import Modal from './Modal';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Modal Component', () => {
  it('should render a Modal component', () => {
    const wrapper = shallow(
      <Modal
        id="testModal"
        modalButtonText="Modal"
        handleSecondaryButton={() => {}}
        secondaryButtonText="Secondary"
        handlePrimaryButton={() => {}}
        primaryButtonText="Primary"
      >
        <p>test</p>
      </Modal>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Modal component with styled buttons', () => {
    const wrapper = shallow(
      <Modal
        id="testModal2"
        modalButtonText="Modal"
        handleSecondaryButton={() => {}}
        secondaryButtonText="Secondary"
        handlePrimaryButton={() => {}}
        primaryButtonText="Primary"
        modalButtonClassName="btn-primary"
        secondaryButtonClassName="btn-secondary"
        primaryButtonClassName="btn-primary"
      >
        <p>test</p>
      </Modal>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Modal component with a header', () => {
    const wrapper = shallow(
      <Modal
        id="testModal3"
        modalButtonText="Modal"
        handleSecondaryButton={() => {}}
        secondaryButtonText="Secondary"
        handlePrimaryButton={() => {}}
        primaryButtonText="Primary"
        hasHeader={true}
        title="Header title"
      >
        <p>test</p>
      </Modal>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
