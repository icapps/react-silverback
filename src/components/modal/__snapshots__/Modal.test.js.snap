// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should render a Modal component 1`] = `ShallowWrapper {}`;

exports[`Modal Component should render a Modal component with a header 1`] = `ShallowWrapper {}`;

exports[`Modal Component should render a Modal component with styled buttons 1`] = `ShallowWrapper {}`;
